### PR TITLE
#139 Fission down

### LINKED ISSUE #139: CLI Command: fission down
# Summary

## Problem

No way to download files through Fission without resorting to IPFS directly

## Solution

Implement the opposite of `fission up`... `fission down`! Simply an alias for `ipfs cat`.

### PR DESCRIPTION
**Pull request by bechurch**
A similar PR may already be submitted!
Please search among the [Pull request](../) before creating one.

Thanks for submitting a pull request! Please provide enough information so that others can review your pull request:

For more information, see the `CONTRIBUTING` guide.

**Summary**

This PR fixes/implements the following **bugs/features**

* [x] Adds a `down` command to download objects from ipfs
* [x] Provide a waiting message during down

Explain the **motivation** for making this change. What existing problem does the pull request solve?

Exposes more ipfs functionality through fission for a better user experience.

**Closing issues**

closes #139

### CODE REVIEW

**Comment by expede** [comment]
👀

File: library/Fission/CLI/Command/Down.hs [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        => cfg
+        -> CommandM (m ())
+command cfg =
+  addCommand
+    "down"
+    "pull a ipfs or ipns object down to your system"
+    (\cid -> runRIO cfg $ down cid)
+    (strArgument (
+      metavar "ContentID" <> help "The CID of the IPFS object you want to download"
+    ))
```
Comment (expede) [inline library/Fission/CLI/Command/Down.hs]: I would rephrase with a `$`:

```haskell
command cfg =
  addCommand
    "down"
    "pull a ipfs or ipns object down to your system"
    (\cid -> runRIO cfg $ down cid)
    (strArgument $ metavar "ContentID" <> help "The CID of the IPFS object you want to download")
```

File: library/Fission/CLI/Command/Down.hs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  case getResult of
+    Right _ok -> do
+      let
+        rawCID = IPFS.unaddress cid
+        successMessage = rawCID <> " Successfully downloaded!"
+
+      CLI.Success.putOk successMessage
+
+    Left  err ->
+      CLI.Error.put err "Oh no! The download failed unexpectedly"
```
Comment (expede) [inline library/Fission/CLI/Command/Down.hs]: This feels a bit wordy. Have you considered:

```haskell
--   Destructuring
--   vvvvvvvvvvvvvv
down cid@(CID hash) = do
  cfg <- ask
  getResult <- liftIO
              . CLI.Wait.waitFor "Retrieving Object..."
              . runRIO cfg
              $ IPFS.getContent cid

  case getResult of --   Use it here vvvv
    Right _ok -> CLI.Success.putOk $ hash <> " Successfully downloaded!"
    Left  err -> CLI.Error.put err "Oh no! The download failed unexpectedly"
```

or even

```haskell
down cid@(CID hash) = do
  cfg <- ask
  liftIO (CLI.Wait.waitFor "Retrieving Object..." $ runRIO cfg $ IPFS.getContent cid) >>= \case
    Right _ok -> CLI.Success.putOk $ hash <> " Successfully downloaded!"
    Left  err -> CLI.Error.put err "Oh no! The download failed unexpectedly"
```

I'm going to get rid of that `liftIO` in a quick PR in a second here...

Comment (bechurch) [reply library/Fission/CLI/Command/Down.hs]: I'm tempted to go with
```haskell
down cid@(IPFS.CID hash) = cid
  & IPFS.getContent
  & CLI.Wait.waitFor "Retrieving Object..."
  >>= \case
    Right _ok ->
      CLI.Success.putOk $ hash <> " Successfully downloaded!"
    Left  err ->
      CLI.Error.put err "Oh no! The download failed unexpectedly"
```
😛
however I think this is better
```haskell
down cid@(IPFS.CID hash) = do
  getResult <- CLI.Wait.waitFor "Retrieving Object..."
              $ IPFS.getContent cid

  case getResult of
    Right _ok ->
      CLI.Success.putOk $ hash <> " Successfully downloaded!"
    Left  err ->
      CLI.Error.put err "Oh no! The download failed unexpectedly"
```
as it clearly separates into "during computation" and after

Comment (bechurch) [reply library/Fission/CLI/Command/Down.hs]: actually... let me know your thoughts on my obvious elixir style!

**Comment by bechurch** [comment]
@expede back up! also took the time to remove some redundant constraints and rename our ipfs `get` to `getFile` and `getContent` to `getDirectoryOrFile` im not stoked on the names so let me know if you can think of something more descriptive

**Comment by expede** [comment]
@bechurch sorry I got all distracted when @bmann and @dholms walked into the coworking space. I made those changed that I alluded to in #150. Lemme know how you feel about that version / I can walk you through those changes when you get here if you like.

File: library/Fission/CLI/Command/Down.hs (hunk 2 above)
Comment (expede) [reply library/Fission/CLI/Command/Down.hs]: @bechurch yeah, it's actually harder to read IMO. You're working in different levels, it's noisier, and so on. I totally get that you're used to thinking in pipes, but you won't be able to rely on this style in the long term — there's lots of code that will have a higher cognitive overhead by trying to pipe like this.

I've been thinking about this a bit. If we as a team decide to go totally off pf the idiomatic style and use pipes, we should consider the Elm style. It's less elegant and noisier, but is I think the most readable pipe style for this language family:

```haskell
down cid@(IPFS.CID hash) = 
  cid
    |> IPFS.getContent
    |> CLI.Wait.waitFor "Retrieving Object..."
    |> bind \case
      Right _ok ->
        CLI.Success.putOk $ hash <> " Successfully downloaded!"
      Left  err ->
        CLI.Error.put err "Oh no! The download failed unexpectedly"
```

**Comment by bechurch** [comment]
Merged this code in #150
